### PR TITLE
[DeepSeek-V4] Implement core primitives (RoPE, GroupedLinear)

### PR DESCRIPTION
**Pull request by parambole**
### Description

Implement foundational architectural core primitives required for DeepSeek-V4 integration into MaxText. These modules serve as the underlying foundation for the V4 architecture.

-   `DeepSeekGroupedLinear`: Block-diagonal grouped linear projection layer. This module enables parallel, memory-efficient group projections natively via `jax.lax.dot_general` / `einsum` broadcasting (`[B, S, hc_mult, D] -> [B, S, D]`).
-   `DeepSeekV4RotaryEmbedding`: Interleaved partial rotary positional embedding that pairs consecutive even/odd channels (as opposed to standard half-split pairs).
-   `Unit Tests`: Added unit tests (`tests/unit/deepseek_v4_vs_reference_test.py`) validating exact numerical parity against PyTorch reference implementations at `atol=1e-5, rtol=1e-5`.

### Tests

Tested on CPU

bash
```
pytest  tests/unit/deepseek_v4_vs_reference_test.py::DeepSeekV4RotaryEmbeddingTest  tests/unit/deepseek_v4_vs_reference_test.py::DeepSeekV4GroupedLinearTest

========================= 3 passed, 1 warning in 2.52s =========================

tests/unit/deepseek_v4_vs_reference_test.py ...                          [100%]
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/layers/embeddings.py [hunk 1]
```diff
@@ -1800,3 +1801,115 @@ def qwen3_omni_mrope_embedding_as_linen(
...
+
+    # Compute base inverse frequencies for half of self.dim (dim // 2 unique theta angles).
+    # Adjacent channels share the same base frequency, matching the reference sequence.
+    half_dim = self.dim // 2
+    fraction = 2 * jnp.arange(0, half_dim, dtype=jnp.float32) / self.dim
+    self.inv_freq = 1.0 / (self.rope_theta**fraction)
+
+  def __call__(self, x: jnp.ndarray, position_ids: jnp.ndarray) -> tuple[jnp.ndarray, jnp.ndarray]:
+    # position_ids: [B, S]
+    # Expand inverse frequencies for broadcasting: [1, 1, dim/2]
```
Comment (dipakg-lang) [inline src/maxtext/layers/embeddings.py]: Would it make sense to replace these three lines with?

freqs = jnp.einsum('bs,d->bsd', position_ids.astype(jnp.float32), self.inv_freq)

**Review by dipakg-lang — COMMENTED** [review]
Please ignore, I setup the github account, was testing if I can indeed comment on the PR.

File: src/maxtext/layers/embeddings.py [hunk 2]
```diff
@@ -1800,3 +1801,115 @@ def qwen3_omni_mrope_embedding_as_linen(
...
+    inv_freq_expanded = self.inv_freq[jnp.newaxis, jnp.newaxis, :]
+
+    # Expand position IDs: [B, S, 1]
+    position_ids_expanded = position_ids[:, :, jnp.newaxis].astype(jnp.float32)
+
+    # Compute outer product of positions and frequencies: [B, S, dim/2]
+    freqs = position_ids_expanded * inv_freq_expanded
+
```
[the remark this answers is not in the document]
Comment (dipakg-lang) [reply src/maxtext/layers/embeddings.py:1847]: +1.

File: src/maxtext/layers/embeddings.py [hunk 3]
```diff
@@ -1800,3 +1801,115 @@ def qwen3_omni_mrope_embedding_as_linen(
...
+  # Interleave consecutive components: [-x2_0, x1_0, -x2_1, x1_1, ...]
+  stacked = jnp.stack((-x2, x1), axis=-1)  # [B, S, H, D_rope/2, 2]
+  return stacked.reshape(x.shape)  # [B, S, H, D_rope]
+
+
+def apply_rotary_pos_emb(
+    x: jax.Array,
+    cos: jax.Array,
+    sin: jax.Array,
+    unsqueeze_dim: int = 2,
```
Comment (dipakg-lang) [inline src/maxtext/layers/embeddings.py]: This assumes your input x will always strictly be 4D: [Batch, Sequence, Head, Dimension], I know this is very sensible assumption, but If you ever pass a single sequence without a batch dimension [S, H, D], or a 5D tensor, axis=2 will unsqueeze the wrong dimension and crash the broadcast.

Please see if we can do something like this :

# Instead of relying on a hardcoded unsqueeze_dim
    head_dim_axis = x.ndim - 2
    cos = jnp.expand_dims(cos, axis=head_dim_axis)
    sin = jnp.expand_dims(sin, axis=head_dim_axis)

File: src/maxtext/layers/linears.py [hunk 4]
```diff
@@ -567,3 +567,66 @@ def mlp_block(
...
+    """Projects segmented groups from the input tensor using block weight matrices.
+
+    Args:
+      x: Input tensor of shape [..., n_groups, in_features_per_group]
+
+    Returns:
+      Projected tensor of shape [..., n_groups, out_features_per_group]
+    """
+    x = jnp.asarray(x, self.dtype)
+    weight = jnp.asarray(self.weight[...], self.dtype)
```
Comment (dipakg-lang) [inline src/maxtext/layers/linears.py]: change this to weight = jnp.asarray(self.weight.value, self.dtype)? for consistency with normalizations.py?

File: tests/unit/deepseek_v4_vs_reference_test.py [hunk 5]
```diff
@@ -0,0 +1,1046 @@
+# pylint: skip-file
```
Comment (shuningjin) [inline tests/unit/deepseek_v4_vs_reference_test.py]: skip this test as it requires torch, in `pytest.ini`
https://github.com/AI-Hypercomputer/maxtext/blob/d324c49bb912eca4bf5eafa6a07143a8c687f12d/pytest.ini#L8

File: src/maxtext/layers/normalizations.py [hunk 6]
```diff
@@ -240,3 +240,62 @@ def l2norm(x: Array, dim: int = -1, eps: float = 1e-6) -> Array:
...
+      dtype: Any = jnp.float32,
+      weight_dtype: Any = jnp.float32,
+  ):
+    self.hidden_size = hidden_size
+    self.eps = eps
+    self.dtype = dtype
+    self.weight_dtype = weight_dtype
+
+    # Initialize learnable scale weight to ones matching T5LayerNorm behavior
+    self.weight = nnx.Param(jnp.ones((hidden_size,), dtype=weight_dtype))
```
[the remark this answers is not in the document]
Comment (shuningjin) [reply src/maxtext/layers/normalizations.py]: +1

File: src/maxtext/layers/normalizations.py [hunk 7]
```diff
@@ -240,3 +240,62 @@ def l2norm(x: Array, dim: int = -1, eps: float = 1e-6) -> Array:
...
+    # Apply reciprocal square root with epsilon offset
+    normalized = x_f32 * lax.rsqrt(variance + self.eps)  # [B, S, D]
+
+    # Cast back to active precision and apply scaling weight
+    y = jnp.asarray(normalized, self.dtype)  # [B, S, D]
+    weight = jnp.asarray(self.weight.get_value(), self.dtype)  # [D]
+    return y * weight  # [B, S, D]
+
+
+class DeepSeekV4UnweightedRMSNorm(nnx.Module):
```
Comment (shuningjin) [inline src/maxtext/layers/normalizations.py]: This perhaps can be replaced with `RMSNorm(num_features=hidden_size, epsilon=eps, dtype=dtype, with_scale=False)`

https://github.com/AI-Hypercomputer/maxtext/blob/e81f52d7d1c4d5351d3b48d06f3e11dd03afcdad/src/maxtext/layers/normalizations.py#L33

File: src/maxtext/layers/normalizations.py [hunk 8]
```diff
@@ -240,3 +240,62 @@ def l2norm(x: Array, dim: int = -1, eps: float = 1e-6) -> Array:
     scale_init=linen_initializers.zeros,
     scale_offset=1.0,
 )
+
+
+class DeepSeekV4RMSNorm(nnx.Module):
```
Comment (shuningjin) [inline src/maxtext/layers/normalizations.py]: This perhaps can be replaced with `RMSNorm(num_features=hidden_size, epsilon=eps, dtype=dtype)`

https://github.com/AI-Hypercomputer/maxtext/blob/e81f52d7d1c4d5351d3b48d06f3e11dd03afcdad/src/maxtext/layers/normalizations.py#L33

Note `y * weight` equivalent to `jnp.einsum("...k,k->...k", y, effective_scale)`

File: src/maxtext/layers/linears.py [hunk 9]
```diff
@@ -567,3 +567,66 @@ def mlp_block(
...
+
+    # Execute parallel group projection via optimized einsum broadcasting.
+    # x: [..., g, i]
+    # weight: [g, i, o]
+    # output: [..., g, o]
+    return jnp.einsum("...gi,gio->...go", x, weight)
```
Comment (shuningjin) [inline src/maxtext/layers/linears.py]: jnp.einsum(..., precision=self.config.matmul_precision)

**Review by shuningjin — COMMENTED** [review]
Thanks! Probably can reuse existing rms-norm. Might be good to add sharding for nnx.Param, or at least add TODO in comments.

File: src/maxtext/layers/linears.py [hunk 10]
```diff
@@ -567,3 +567,66 @@ def mlp_block(
...
+    kernel_shape = (n_groups, in_features_per_group, self.out_features_per_group)
+    self.weight = nnx.Param(
+        kernel_init(
+            rngs.params(),
+            kernel_shape,
+            self.weight_dtype,
+            in_axis=1,
+            out_axis=2,
+        )
+    )
```
[the remark this answers is not in the document]
Comment (shuningjin) [reply src/maxtext/layers/linears.py]: +1. Add `sharding=kernel_axes` to `nnx.Param`, similar for other places. At least add `#TODO`.

File: src/maxtext/layers/embeddings.py (hunk 1 above)
Comment (dipakg-lang) [reply src/maxtext/layers/embeddings.py]: .

File: src/maxtext/layers/embeddings.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (dipakg-lang) [reply src/maxtext/layers/embeddings.py:1847]: .

File: src/maxtext/layers/embeddings.py (hunk 3 above)
Comment (dipakg-lang) [reply src/maxtext/layers/embeddings.py]: .

File: src/maxtext/layers/linears.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (entrpn) [reply src/maxtext/layers/linears.py]: Do we have an idea of what the kernel_axes names should be? if not, the grouped projections might not be sharded efficiently and if that's the case, lets add a comment with a todo for optimizing this.

File: src/maxtext/layers/embeddings.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (parambole) [reply src/maxtext/layers/embeddings.py:1847]: Great suggestion. I have updated DeepSeekV4RotaryEmbedding.__call__ to maintain both sinusoids strictly in jnp.float32. This ensures numerical stability up to the final rotation step and locks in exact parity with my PyTorch golden reference. Resolved.

File: src/maxtext/layers/linears.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (parambole) [reply src/maxtext/layers/linears.py]: Thanks for highlighting this. I have added kernel_axes parameterization to the module constructor and passed sharding=kernel_axes into nnx.Param.

File: src/maxtext/layers/embeddings.py (hunk 1 above)
Comment (parambole) [reply src/maxtext/layers/embeddings.py]: Thanks for the suggestion! I kept the outer product broadcasting (position_ids_expanded * inv_freq_expanded) because XLA automatically fuses the elementwise multiplication with jnp.cos and jnp.sin during compilation, while keeping the precision strictly in float32 for numerical rigor.

File: src/maxtext/layers/embeddings.py (hunk 3 above)
Comment (parambole) [reply src/maxtext/layers/embeddings.py]: Spot on catch! I have completely replaced the hardcoded unsqueeze_dim=2 parameter with dynamic calculation (head_dim_axis = x.ndim - 2) inside apply_rotary_pos_emb. My RoPE expansion is now rank invariant and robust to unbatched sequences.

File: tests/unit/deepseek_v4_vs_reference_test.py (hunk 5 above)
Comment (parambole) [reply tests/unit/deepseek_v4_vs_reference_test.py]: Added it to the file.

File: src/maxtext/layers/linears.py (hunk 9 above)
Comment (parambole) [reply src/maxtext/layers/linears.py]: I have added the precision.

File: src/maxtext/layers/normalizations.py (hunk 8 above)
Comment (parambole) [reply src/maxtext/layers/normalizations.py]: Absolutely. I have completely removed the custom RMS norm class and transitioned my test suite to directly instantiate native RMSNorm(num_features=512, epsilon=1e-6, with_scale=False, rngs=None). The parity tests pass cleanly at atol=1e-5

File: src/maxtext/layers/normalizations.py (hunk 7 above)
Comment (parambole) [reply src/maxtext/layers/normalizations.py]: Done!

File: src/maxtext/layers/linears.py (hunk 4 above)
Comment (parambole) [reply src/maxtext/layers/linears.py]: I have refactored self.weight to cleanly reference self.kernel[...], adhering perfectly to Flax NNX Variable array indexing paradigms. Resolved.

**Review by entrpn — APPROVED** [review]

**Review by shuningjin — APPROVED** [review]
Thank you! LGTM!

**Review by dipakg-lang — COMMENTED** [review]
LGTM!
